pypose/module/pnp.py

```diff
@@ -0,0 +1,308 @@
+            \sum^4_{j=1}{\alpha_{ij}f_yy^c_j + \alpha_{ij}(v_0 - v_i)z^c_j} &= 0
+        \end{aligned}
+
+    Assume :math:`\mathbf{x} = \begin{bmatrix} c^{c^T}_1 & c^{c^T}_2 & c^{c^T}_3 &
+    c^{c^T}_4 \end{bmatrix}^T`, then the two equations form a system :math:`Mx = 0`
+    considering all of the :math:`n` points. Its solution can be expressed as
+
+    .. math::
+        \begin{aligned}
+            x &= \sum^4_{i=1}{\beta_iv_i},
+        \end{aligned}
+
+    where :math:`v_i` is the null vectors of matrix :math:`M^T M` corresponding to its
+    least 4 eigenvalues.
+
+    The final step involves calculating the coefficients :math:`\beta_i`. Optionally, the
+    Gauss-Newton algorithm can be used to refine the solution of :math:`\beta_i`.
+
+    Example:
+        >>> import torch, pypose as pp
+        >>> f, (H, W) = 2, (9, 9) # focal length and image height, width
+        >>> intrinsics = torch.tensor([[f, 0, H / 2],
+        ...                            [0, f, W / 2],
+        ...                            [0, 0,   1  ]])
+        >>> object = torch.tensor([[2., 0., 2.],
+        ...                        [1., 0., 2.],
+        ...                        [0., 1., 1.],
+        ...                        [0., 0., 1.],
+        ...                        [1., 0., 1.],
+        ...                        [5., 5., 3.]])
+        >>> pixels = pp.point2pixel(object, intrinsics)
+        >>> pose = pp.SE3([ 0., -8,  0.,  0., -0.3827,  0.,  0.9239])
+        >>> points = pose.Inv() @ object
+        ...
+        >>> epnp = pp.module.EPnP(intrinsics)
+        >>> pose = epnp(points, pixels)
+        SE3Type LieTensor:
+        LieTensor([ 3.9816e-05, -8.0000e+00,  5.8174e-05, -3.3186e-06, -3.8271e-01,
+                    3.6321e-06,  9.2387e-01])
+
+    Warning:
+        Currently this module only supports batched rectified camera intrinsics, which can
+        be defined in the form:
+
+        .. math::
+            K = \begin{pmatrix}
+                    f_x &   0 & c_x \\
+                    0   & f_y & c_y \\
+                    0   &   0 &   1
+                \end{pmatrix}
+
+        The full form of camera intrinsics will be supported in a future release.
+
+    Note:
+        The implementation is based on the paper
+
+        * Francesc Moreno-Noguer, Vincent Lepetit, and Pascal Fua, `EPnP: An Accurate O(n)
+          Solution to the PnP Problem <https://doi.org/10.1007/s11263-008-0152-6>`_,
+          International Journal of Computer Vision (IJCV), 2009.
+    '''
+    def __init__(self, intrinsics=None, refine=True):
+        super().__init__()
+        self.refine = refine
+        self.solver = LSTSQ()
+        if intrinsics is not None:
+            self.register_buffer('intrinsics', intrinsics)
+
+    def forward(self, points, pixels, intrinsics=None):
+        r'''
+        Args:
+            points (``torch.Tensor``): 3D object points in the world coordinates.
+                Shape (..., n, 3)
```

Comment:
In the other comments, "N" is used instead of "n". Consider keeping it consistent

Reply:
Fixed in commit below